src/gwt/src/org/rstudio/studio/client/workbench/ui/PaneManager.java

```diff
@@ -582,55 +583,103 @@ public void onFocusCenterSeparator()
    @Handler
    public void onFocusNextPane()
    {
-      // get the name of the currently focused pane
       LogicalWindow currentFocus = getActiveLogicalWindow();
-      String nextFocusName = null;
-      if (currentFocus != null)
+      if (currentFocus == null)
+         return;
+
+      focusWindow(getAdjacentWindow(currentFocus, false /* before */));
+   }
+
+   @Handler
+   public void onFocusPreviousPane()
+   {
+      LogicalWindow currentFocus = getActiveLogicalWindow();
+      if (currentFocus == null)
+         return;
+      focusWindow(getAdjacentWindow(currentFocus, true /* before */));
+   }
+
+   @Handler
+    public void onNewSourceColumn()
+    {
+       if (!userPrefs_.allowSourceColumns().getValue())
+          pGlobalDisplay_.get().showMessage(GlobalDisplay.MSG_INFO, "Cannot Add Column",
+             "Allow Source Columns preference is disabled.");
+       else if (additionalSourceCount_ == MAX_COLUMN_COUNT)
```

Comment:
Nit: although it's a constant now, if we ever make max column count configurable then maybe want this to be a `>=` comparison just to be safe in case where max changes to be smaller than current number of columns?

Reply:
Good idea, done!